File: pdui-express/lib/core/PDUIGenericFn.ts
import { PDUIFn } from "../models/PDUIFn";
import { type PBFn, type PBFnExpression } from "../proto-out/pdui-proto-out.ts";
import { parseBody } from "./PDUIFnBodyParser";
import { parseParams } from "./PDUIFnParamParser";

export class PDUIGenericFn extends PDUIFn {
    constructor({
        name,
        params,
        body,
    }: {
        name?: string;
        params: { [key: string]: number | string | boolean };
        body: () => PDUIFn[];
    }) {
        super({
            identity: "genericFn",
            name,
            params: parseParams(name, params),
            body: parseBody(name, body),
        });
    }

    override toPB(): PBFn {
        return { name: this.name, params: this.params, body: this.body };
    }

    override toFnExpression(): PBFnExpression {
        return {
            identity: "genericFn",
            payload: {
                oneofKind: "genericFn",
                genericFn: {
                    name: this.name,
                    params: this.params,
                    body: this.body,
                },
            },
        };
    }
}

File: pdui_flutter/lib/core/pdui_generic_fn.dart
import '../models/pdui_fn.dart';
import '../proto_out/pdui-proto-out.pb.dart';
import 'pdui_expression_resolver.dart';

class PduiGenericFn extends PduiFn {
  PBFnExpression pbFnExpression;

  PduiGenericFn(this.pbFnExpression) : super("generic");

  @override
  resolve() {
    pbFnExpression.genericFn.body.forEach((key, body) {
      PduiFnExpressionResolver.parse(body);
    });
  }
}
